Source/AwsCommonRuntimeKit/crt/Future.swift

```diff
@@ -0,0 +1,78 @@
+//  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//  SPDX-License-Identifier: Apache-2.0.
+
+import Foundation
+
+public final class Future<Value> {
+    public typealias FutureResult = Result<Value, Error>
+
+    private var _value: FutureResult? //nil when pending
+
+    private var waiter = DispatchSemaphore(value: 0)
+
+    private var _observers: [((FutureResult) -> Void)] = []
```

Comment:
All accesses to this array need to be protected by a lock/mutex or something. You could use a mutex for all access to `_value` and `_observers` instead of getting atomics involved for simplicity if you want.